Index: libs/WindowManager/Shell/src/com/android/wm/shell/windowdecor/DragResizeInputListener.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libs/WindowManager/Shell/src/com/android/wm/shell/windowdecor/DragResizeInputListener.java b/libs/WindowManager/Shell/src/com/android/wm/shell/windowdecor/DragResizeInputListener.java
--- a/libs/WindowManager/Shell/src/com/android/wm/shell/windowdecor/DragResizeInputListener.java	(revision f95ddb5766067a975d86f34a9b6b265c0033e2e2)
+++ b/libs/WindowManager/Shell/src/com/android/wm/shell/windowdecor/DragResizeInputListener.java	(revision a1df0b6e053eb19e2af4b17d7724df9ef080b2f4)
@@ -17,6 +17,7 @@
 package com.android.wm.shell.windowdecor;
 
 import static android.view.WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE;
+import static android.view.WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE;
 import static android.view.WindowManager.LayoutParams.PRIVATE_FLAG_TRUSTED_OVERLAY;
 import static android.view.WindowManager.LayoutParams.TYPE_APPLICATION;
 
@@ -90,7 +91,7 @@
                     mDecorationSurface,
                     mFakeWindow,
                     null /* hostInputToken */,
-                    FLAG_NOT_FOCUSABLE,
+                    FLAG_NOT_TOUCHABLE,
                     PRIVATE_FLAG_TRUSTED_OVERLAY,
                     TYPE_APPLICATION,
                     mFocusGrantToken,
@@ -151,7 +152,7 @@
                     mInputChannel.getToken(),
                     mDisplayId,
                     mDecorationSurface,
-                    FLAG_NOT_FOCUSABLE,
+                    FLAG_NOT_TOUCHABLE,
                     PRIVATE_FLAG_TRUSTED_OVERLAY,
                     touchRegion);
         } catch (RemoteException e) {
